File: scripts/import-photos.py
#!/usr/bin/env python3
"""
Import photos from naturephotos folder to public/images/nature and generate parks.json
"""
import os
import json
import shutil
from pathlib import Path

# Configuration
SOURCE_DIR = Path(__file__).parent.parent / 'naturephotos'
DEST_IMAGE_DIR = Path(__file__).parent.parent / 'public' / 'images' / 'nature'
OUTPUT_JSON = Path(__file__).parent.parent / 'src' / 'data' / 'parks.json'

# Supported image extensions
IMAGE_EXTENSIONS = {'.jpg', '.jpeg', '.png', '.webp', '.gif'}

def name_to_id(name):
    """Convert park name to URL-friendly ID (lowercase, hyphens)"""
    import re
    # Convert to lowercase, replace non-alphanumeric with hyphens
    park_id = re.sub(r'[^a-z0-9]+', '-', name.lower())
    # Remove leading/trailing hyphens
    return park_id.strip('-')

def parse_credits(credits_path):
    """Parse credits.txt file and return a dict mapping filename to credit info"""
    credits = {}
    
    if not credits_path.exists():
        print(f"  ⚠️  No credits.txt found in {credits_path.parent}")
        return credits
    
    try:
        content = credits_path.read_text(encoding='utf-8')
        sections = [s.strip() for s in content.split('--------------------') if s.strip()]
        
        for section in sections:
            lines = section.strip().split('\n')
            filename = None
            author = None
            license_text = None
            source = None
            
            for line in lines:
                line = line.strip()
                if line.startswith('File:'):
                    filename = line.replace('File:', '').strip()
                elif line.startswith('Author:'):
                    author = line.replace('Author:', '').strip()
                elif line.startswith('License:'):
                    license_text = line.replace('License:', '').strip()
                elif line.startswith('Source:'):
                    source = line.replace('Source:', '').strip()
            
            if filename:
                credits[filename] = {
                    'author': author or 'Unknown',
                    'license': license_text or 'Unknown',
                    'source': source or None
                }
    except Exception as e:
        print(f"  ⚠️  Error parsing credits.txt: {e}")
    
    return credits

def get_priority(filename):
    """Determine priority based on filename"""
    lower_name = filename.lower()
    if 'hero' in lower_name or 'cover' in lower_name:
        return 'hero'
    return 'gallery'

def process_park_folder(park_name, park_path):
    """Process a single park folder"""
    park_id = name_to_id(park_name)
    print(f"\n📁 Processing: {park_name} (ID: {park_id})")
    
    # Create destination directory
    dest_dir = DEST_IMAGE_DIR / park_id
    dest_dir.mkdir(parents=True, exist_ok=True)
    print(f"  ✓ Created/verified directory: {dest_dir}")
    
    # Parse credits
    credits_path = park_path / 'credits.txt'
    credits = parse_credits(credits_path)
    
    # Get all image files
    image_files = [
        f for f in os.listdir(park_path)
        if Path(f).suffix.lower() in IMAGE_EXTENSIONS
    ]
    
    if not image_files:
        print(f"  ⚠️  No images found in {park_name}")
        return None
    
    print(f"  📸 Found {len(image_files)} images")
    
    # Process each image
    images = []
    for image_file in image_files:
        src_path = park_path / image_file
        dest_path = dest_dir / image_file
        
        # Move file (or copy if it already exists in destination)
        if not dest_path.exists():
            shutil.copy2(src_path, dest_path)
            print(f"  ✓ Moved: {image_file}")
        else:
            print(f"  ⊙ Skipped (exists): {image_file}")
        
        # Get credit info
        credit_info = credits.get(image_file, {
            'author': 'Unknown',
            'license': 'Unknown',
            'source': None
        })
        
        # Determine priority
        priority = get_priority(image_file)
        
        # Build image object
        image_obj = {
            'src': f'/images/nature/{park_id}/{image_file}',
            'credit': {
                'author': credit_info['author'],
                'license': credit_info['license']
            },
            'priority': priority
        }
        
        if credit_info.get('source'):
            image_obj['credit']['source'] = credit_info['source']
        
        images.append(image_obj)
    
    return {
        'parkId': park_id,
        'title': park_name,
        'images': images
    }

def main():
    """Main function"""
    print('🚀 Starting photo import process...\n')
    print(f'Source: {SOURCE_DIR}')
    print(f'Destination: {DEST_IMAGE_DIR}')
    print(f'Output: {OUTPUT_JSON}')
    
    # Check if source directory exists
    if not SOURCE_DIR.exists():
        print(f'❌ Error: Source directory not found: {SOURCE_DIR}')
        return 1
    
    # Create destination directories
    DEST_IMAGE_DIR.mkdir(parents=True, exist_ok=True)
    OUTPUT_JSON.parent.mkdir(parents=True, exist_ok=True)
    print(f'✓ Created/verified directories')
    
    # Read existing parks.json if it exists
    existing_parks = {}
    if OUTPUT_JSON.exists():
        try:
            with open(OUTPUT_JSON, 'r', encoding='utf-8') as f:
                existing = json.load(f)
            # Convert array to dict for easier lookup
            if isinstance(existing, list):
                for park in existing:
                    existing_parks[park['parkId']] = park
            elif isinstance(existing, dict):
                existing_parks = existing
            print(f'\n📖 Found existing parks.json with {len(existing_parks)} parks')
        except Exception as e:
            print(f'⚠️  Could not parse existing parks.json: {e}')
    
    # Get all park folders
    park_folders = [
        d for d in SOURCE_DIR.iterdir()
        if d.is_dir() and not d.name.startswith('.')
    ]
    
    if not park_folders:
        print(f'❌ No park folders found in {SOURCE_DIR}')
        return 1
    
    print(f'\n📂 Found {len(park_folders)} park folders\n')
    
    # Process each park
    parks = []
    for park_path in sorted(park_folders):
        park_name = park_path.name
        park_data = process_park_folder(park_name, park_path)
        
        if park_data:
            # Merge with existing data if available
            if park_data['parkId'] in existing_parks:
                print(f"  🔄 Merging with existing data for {park_data['parkId']}")
                existing = existing_parks[park_data['parkId']]
                # Combine images, avoiding duplicates
                existing_srcs = {img['src'] for img in existing['images']}
                for img in park_data['images']:
                    if img['src'] not in existing_srcs:
                        existing['images'].append(img)
                # Update title if needed
                if existing['title'] != park_data['title']:
                    print(f"  📝 Updating title: \"{existing['title']}\" → \"{park_data['title']}\"")
                    existing['title'] = park_data['title']
                parks.append(existing)
            else:
                parks.append(park_data)
    
    # Sort parks by title
    parks.sort(key=lambda x: x['title'])
    
    # Write output JSON
    with open(OUTPUT_JSON, 'w', encoding='utf-8') as f:
        json.dump(parks, f, indent=2, ensure_ascii=False)
    
    print(f'\n✅ Success! Processed {len(parks)} parks')
    print(f'📝 Output written to: {OUTPUT_JSON}')
    print(f'\n📊 Summary:')
    for park in parks:
        hero_count = sum(1 for img in park['images'] if img['priority'] == 'hero')
        gallery_count = sum(1 for img in park['images'] if img['priority'] == 'gallery')
        print(f"  • {park['title']}: {len(park['images'])} images ({hero_count} hero, {gallery_count} gallery)")
    
    print(f'\n🎉 Done!')
    return 0

if __name__ == '__main__':
    exit(main())

File: scripts/import-photos.js
const fs = require('fs');
const path = require('path');

// Configuration
const SOURCE_DIR = path.join(__dirname, '..', 'naturephotos');
const DEST_IMAGE_DIR = path.join(__dirname, '..', 'public', 'images', 'nature');
const OUTPUT_JSON = path.join(__dirname, '..', 'src', 'data', 'parks.json');

// Supported image extensions
const IMAGE_EXTENSIONS = ['.jpg', '.jpeg', '.png', '.webp', '.gif'];

/**
 * Convert park name to URL-friendly ID (lowercase, hyphens)
 */
function nameToId(name) {
  return name
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, '-')
    .replace(/^-|-$/g, '');
}

/**
 * Parse credits.txt file
 * Returns an object mapping filename to credit info
 */
function parseCredits(creditsPath) {
  const credits = {};
  
  if (!fs.existsSync(creditsPath)) {
    console.warn(`  ⚠️  No credits.txt found in ${creditsPath}`);
    return credits;
  }
  
  const content = fs.readFileSync(creditsPath, 'utf-8');
  const sections = content.split('--------------------').filter(s => s.trim());
  
  for (const section of sections) {
    const lines = section.trim().split('\n');
    let filename = null;
    let author = null;
    let license = null;
    let source = null;
    
    for (const line of lines) {
      if (line.startsWith('File:')) {
        filename = line.replace(/^File:\s*/, '').trim();
      } else if (line.startsWith('Author:')) {
        author = line.replace(/^Author:\s*/, '').trim();
      } else if (line.startsWith('License:')) {
        license = line.replace(/^License:\s*/, '').trim();
      } else if (line.startsWith('Source:')) {
        source = line.replace(/^Source:\s*/, '').trim();
      }
    }
    
    if (filename) {
      credits[filename] = {
        author: author || 'Unknown',
        license: license || 'Unknown',
        source: source || null
      };
    }
  }
  
  return credits;
}

/**
 * Determine priority based on filename
 */
function getPriority(filename) {
  const lowerName = filename.toLowerCase();
  if (lowerName.includes('hero') || lowerName.includes('cover')) {
    return 'hero';
  }
  return 'gallery';
}

/**
 * Process a single park folder
 */
function processParkFolder(parkName, parkPath) {
  const parkId = nameToId(parkName);
  console.log(`\n📁 Processing: ${parkName} (ID: ${parkId})`);
  
  // Create destination directory
  const destDir = path.join(DEST_IMAGE_DIR, parkId);
  if (!fs.existsSync(destDir)) {
    fs.mkdirSync(destDir, { recursive: true });
    console.log(`  ✓ Created directory: ${destDir}`);
  }
  
  // Parse credits
  const creditsPath = path.join(parkPath, 'credits.txt');
  const credits = parseCredits(creditsPath);
  
  // Get all image files
  const files = fs.readdirSync(parkPath);
  const imageFiles = files.filter(file => {
    const ext = path.extname(file).toLowerCase();
    return IMAGE_EXTENSIONS.includes(ext);
  });
  
  if (imageFiles.length === 0) {
    console.log(`  ⚠️  No images found in ${parkName}`);
    return null;
  }
  
  console.log(`  📸 Found ${imageFiles.length} images`);
  
  // Process each image
  const images = [];
  for (const imageFile of imageFiles) {
    const srcPath = path.join(parkPath, imageFile);
    const destPath = path.join(destDir, imageFile);
    
    // Move file (or copy if it already exists in destination)
    if (!fs.existsSync(destPath)) {
      fs.copyFileSync(srcPath, destPath);
      console.log(`  ✓ Moved: ${imageFile}`);
    } else {
      console.log(`  ⊙ Skipped (exists): ${imageFile}`);
    }
    
    // Get credit info
    const creditInfo = credits[imageFile] || {
      author: 'Unknown',
      license: 'Unknown',
      source: null
    };
    
    // Determine priority
    const priority = getPriority(imageFile);
    
    // Build image object
    const imageObj = {
      src: `/images/nature/${parkId}/${imageFile}`,
      credit: {
        author: creditInfo.author,
        license: creditInfo.license
      },
      priority: priority
    };
    
    if (creditInfo.source) {
      imageObj.credit.source = creditInfo.source;
    }
    
    images.push(imageObj);
  }
  
  return {
    parkId: parkId,
    title: parkName,
    images: images
  };
}

/**
 * Main function
 */
function main() {
  console.log('🚀 Starting photo import process...\n');
  console.log(`Source: ${SOURCE_DIR}`);
  console.log(`Destination: ${DEST_IMAGE_DIR}`);
  console.log(`Output: ${OUTPUT_JSON}`);
  
  // Check if source directory exists
  if (!fs.existsSync(SOURCE_DIR)) {
    console.error(`❌ Error: Source directory not found: ${SOURCE_DIR}`);
    process.exit(1);
  }
  
  // Create destination directories
  if (!fs.existsSync(DEST_IMAGE_DIR)) {
    fs.mkdirSync(DEST_IMAGE_DIR, { recursive: true });
    console.log(`✓ Created destination directory`);
  }
  
  if (!fs.existsSync(path.dirname(OUTPUT_JSON))) {
    fs.mkdirSync(path.dirname(OUTPUT_JSON), { recursive: true });
    console.log(`✓ Created data directory`);
  }
  
  // Read existing parks.json if it exists
  let existingParks = {};
  if (fs.existsSync(OUTPUT_JSON)) {
    try {
      const existing = JSON.parse(fs.readFileSync(OUTPUT_JSON, 'utf-8'));
      // Convert array to object for easier lookup
      if (Array.isArray(existing)) {
        existing.forEach(park => {
          existingParks[park.parkId] = park;
        });
      } else if (typeof existing === 'object') {
        existingParks = existing;
      }
      console.log(`\n📖 Found existing parks.json with ${Object.keys(existingParks).length} parks`);
    } catch (e) {
      console.warn(`⚠️  Could not parse existing parks.json: ${e.message}`);
    }
  }
  
  // Get all park folders
  const parkFolders = fs.readdirSync(SOURCE_DIR, { withFileTypes: true })
    .filter(dirent => dirent.isDirectory())
    .map(dirent => dirent.name);
  
  if (parkFolders.length === 0) {
    console.error(`❌ No park folders found in ${SOURCE_DIR}`);
    process.exit(1);
  }
  
  console.log(`\n📂 Found ${parkFolders.length} park folders\n`);
  
  // Process each park
  const parks = [];
  for (const parkName of parkFolders) {
    const parkPath = path.join(SOURCE_DIR, parkName);
    const parkData = processParkFolder(parkName, parkPath);
    
    if (parkData) {
      // Merge with existing data if available
      if (existingParks[parkData.parkId]) {
        console.log(`  🔄 Merging with existing data for ${parkData.parkId}`);
        const existing = existingParks[parkData.parkId];
        // Combine images, avoiding duplicates
        const existingSrcs = new Set(existing.images.map(img => img.src));
        parkData.images.forEach(img => {
          if (!existingSrcs.has(img.src)) {
            existing.images.push(img);
          }
        });
        // Update title if needed
        if (existing.title !== parkData.title) {
          console.log(`  📝 Updating title: "${existing.title}" → "${parkData.title}"`);
        }
        parks.push(existing);
      } else {
        parks.push(parkData);
      }
    }
  }
  
  // Sort parks by title
  parks.sort((a, b) => a.title.localeCompare(b.title));
  
  // Write output JSON
  const output = JSON.stringify(parks, null, 2);
  fs.writeFileSync(OUTPUT_JSON, output, 'utf-8');
  
  console.log(`\n✅ Success! Processed ${parks.length} parks`);
  console.log(`📝 Output written to: ${OUTPUT_JSON}`);
  console.log(`\n📊 Summary:`);
  parks.forEach(park => {
    const heroCount = park.images.filter(img => img.priority === 'hero').length;
    const galleryCount = park.images.filter(img => img.priority === 'gallery').length;
    console.log(`  • ${park.title}: ${park.images.length} images (${heroCount} hero, ${galleryCount} gallery)`);
  });
  
  console.log(`\n🎉 Done!`);
}

// Run the script
if (require.main === module) {
  main();
}

module.exports = { main };

